resolve.js

```diff
@@ -0,0 +1,123 @@
+const assert = require('assert')
+const encoding = require('dat-encoding')
+const https = require('https')
+const NodeCache = require( "node-cache" )
+const querystring = require('querystring')
+const url = require('url')
+
+function Resolver() {
+  this._myCache = new NodeCache()
+}
+
+Resolver.prototype.close = function() {
+  this._myCache.close()
+}
+
+Resolver.prototype.resolve = function(href, opts, cb) {
+  if (typeof opts === 'function') {
+    cb = opts
+    opts = {}
+  }
+
+  if (!href || typeof href !== 'string' || href.length === 0) {
+    cb('Invalid href')
+    return
+  }
+
+  // Resolve key in href
+  try {
+    const key = encoding.encode(encoding.decode(href))
+    cb(null, key)
+  }
+  catch (err) {
+    const parsed_url = url.parse(href)
+    const hostname = parsed_url.hostname || parsed_url.pathname
+
+    this._resolveFromCache(hostname, opts, (err, key) => {
+      if (err) {
+        cb(err)
+      } else {
+        if (key) {
+          cb(null, key)
+        } else {
+          resolveWithDns(hostname, opts, (err, key) => {
+            if (err) {
+              cb(err)
+            } else {
+              cb(null, key)
+            }
+          })
+        }
+      }
+    })
+  }
+}
+
+Resolver.prototype._resolveFromCache = function(hostname, opts, cb) {
+  if (typeof opts === 'function') {
+    cb = opts
+    opts = {}
+  }
+
+  const cache = opts.cache || this._myCache
+
+  cb(null, cache.get(hostname))
+}
+
+const CABAL_KEY_REGEX = /^"?cabalkey=([0-9a-f]{64})"?$/i
+
+function parseKeyFromDns(data) {
+  const answers = data.Answer || []
+  const cabal_answers = answers.filter((answer) => {
+    const matches = CABAL_KEY_REGEX.exec(answer.data)
+
+    if (matches && matches.length > 0) {
+      return true
+    } else {
+      return false
+    }
+  })
+
+  const answer = cabal_answers[0]
+  const key = CABAL_KEY_REGEX.exec(answer.data)[1]
+  return key
+}
+
+function dnsResolver(hostname, cb) {
```

Comment:
Could we use node's `dns` module here?  e.g.

```
$ cat > dns.js
var dns = require('dns')

dns.resolveTxt('beakerbrowser.com', console.log)
^D

$ node dns.js 
null [ [ 'DATKEY=87ed2e3b160f261a032af03921a3bd09227d0a4cde73466c17114816cae43336' ] ]
```

Reply:
I noticed that `dat-dns` chose to use DNS-over-HTTPS as they believed that it offered more confidential lookups, hence using it here. I'm fine with using the node dns if you think that's ok. It's written so that the dns resolver can be easily replaced if someone had concerns one way or the other.

Just let me know what DNS you want and I'll wire it up!

Reply:
Can you say a bit more about this, or link me to what the dat folks said?

Is the idea that hitting your ISP for DNS (probably most people's default) is less desirable than google's servers?

Reply:
Sure. I can't remember exactly where I saw it but I'll track it down!

Reply:
I've dropped the Google DNS and am just using whatever's configured on the system.

Along with this change, I've also dropped the caching and gotten rid of `Cabal.close()` as well.